### PR TITLE
[CDAP-15787] Updated ExcelInputReader for New Validation API

### PR DESCRIPTION
**Pull request by dli357**
Updated ExcelInputReader.java for the new validation API. Was unable to run the unit tests due to an InvalidArtifactException.

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 1]
```diff
@@ -129,12 +138,13 @@ private void init() {
       for (String map : mappings) {
         String[] columns = map.split(":");
         if (CollectionUtils.isNotEmpty(inputColumns) && !inputColumns.contains(columns[0])) {
-          throw new IllegalArgumentException("Column name: " + columns[0] + " in 'Column-Label Mapping' does not " +
-                                               "match the columns in the 'Column To Be Extracted' input text box. " +
-                                               "It has to be one of the columns in 'Column To Be Extracted' " +
-                                               "input text box.");
+          collector.addFailure(
+              String.format("Column: %s in 'Column-Label Mapping' "
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: indentation

Comment (albertshau) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: single quotes around '%s'

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 2]
```diff
@@ -143,14 +153,16 @@ private void init() {
       for (String schema : schemaList) {
         String[] columns = schema.split(":");
         if (CollectionUtils.isNotEmpty(inputColumns) && !inputColumns.contains(columns[0])) {
-          throw new IllegalArgumentException("Column name: " + columns[0] + " in 'Field Name Schema Type Mapping'" +
-                                               " does not match the columns in the 'Column To Be Extracted' input " +
-                                               "text box. It has to be one of the columns in " +
-                                               "'Column To Be Extracted' input text box.");
+          collector.addFailure(
+              String.format("Column: %s in 'Field Name Schema Type Mapping' "
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: single quotes around '%s'

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 3]
```diff
@@ -310,6 +325,15 @@ public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
   }
 
   private void createDatasets(@Nullable PipelineConfigurer pipelineConfigurer, @Nullable BatchSourceContext context) {
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: FYI, this is a bad pattern. A method shouldn't take two arguments, one of which it expects to always be null and the other not.

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 4]
```diff
@@ -310,6 +325,15 @@ public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
...
+    FailureCollector collector;
+    if (pipelineConfigurer != null) {
+      collector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    } else if (context != null) {
+      collector = context.getFailureCollector();
+    } else {
+      throw new IllegalArgumentException("Unable to retrieve failure collector.");
+    }
     try {
       if (!excelInputreaderConfig.containsMacro("errorDatasetName") &&
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java:362]: woah just read how this plugin is handling errors and it is completely wrong. Can you modify the widget .json file for this plugin to set the widget type for errorDatasetName and ifErrorRecord to 'hidden' so nobody tries to use them?

In fact, I think the existence of this entire source is questionable, so I've opened CDAP-15937 to revisit this next release.

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 5]
```diff
@@ -143,14 +153,16 @@ private void init() {
...
+              String.format("Column: %s in 'Field Name Schema Type Mapping' "
+                  + "must be included in 'Column To Be Extracted'", columns[0]), null)
+              .withConfigElement(OUTPUT_SCHEMA, schema);
+        } else {
+          outputSchemaMapping.put(columns[0], columns[1]);
         }
-        outputSchemaMapping.put(columns[0], columns[1]);
       }
     }
+    collector.getOrThrowException();
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: whether to call this or not should be decided by the caller and not placed here at the end of the method

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 6]
```diff
@@ -489,20 +515,25 @@ public ExcelInputReaderConfig() {
...
+            .withConfigProperty(TABLE_EXPIRY_PERIOD);
       }
 
       if (!Strings.isNullOrEmpty(rowsLimit) && !StringUtils.isNumeric(rowsLimit)) {
-        throw new IllegalArgumentException(String.format("Invalid row limit: %s. " +
-                                                           "Numeric value expected.", rowsLimit));
+        collector.addFailure(String.format("Invalid row limit: '%s'.", rowsLimit), "Numeric value expected.")
+            .withConfigProperty(ROWS_LIMIT);
       }
+      collector.getOrThrowException();
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: should not be here, the caller should decide whether to call it. In the case of configurePipeline(), it looks like more things can be validated.

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 7]
```diff
@@ -266,23 +278,26 @@ public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRe
...
-    init();
+    init(collector);
     getOutputSchema();
     pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
   }
 
   @Override
   public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
-    excelInputreaderConfig.validate();
+    excelInputreaderConfig.validate(batchSourceContext.getFailureCollector());
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: call getOrThrowException after this method call

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 8]
```diff
@@ -266,23 +278,26 @@ public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRe
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     super.configurePipeline(pipelineConfigurer);
 
-    excelInputreaderConfig.validate();
+    // Get failure collector for updated validation API
+    FailureCollector collector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    excelInputreaderConfig.validate(collector);
 
     if (Strings.isNullOrEmpty(excelInputreaderConfig.columnList) &&
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: this looks like it belongs in the validate() method

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 1 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 2 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 3 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Yeah I built off of the existing method which already did this, I can go ahead and create overloaded methods instead which are probably better.

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 4 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java:362]: Sounds good, changed to hidden for now

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 5 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 6 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 7 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 8 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 3 above)
Comment (albertshau) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: its ok, you don't have to refactor this, was just mentioning it.

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java [hunk 9]
```diff
@@ -309,7 +333,37 @@ public void prepareRun(BatchSourceContext batchSourceContext) throws Exception {
...
+        collector.addFailure("Error dataset name should not be empty if choosing write to error "
+                               + "dataset for 'On Error' input.", null)
+          .withConfigProperty(IF_ERROR_RECORD).withConfigProperty(ERROR_DATASET_NAME);
+      }
+
+      if (!excelInputreaderConfig.containsMacro("memoryTableName") &&
+        !Strings.isNullOrEmpty(excelInputreaderConfig.memoryTableName)) {
+        pipelineConfigurer.createDataset(excelInputreaderConfig.memoryTableName, KeyValueTable.class);
+      }
+    } catch (Exception e) {
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: don't think anything throws an exception anymore. If not, you should remove this

**Review by albertshau — APPROVED** [review]
minor comment otherwise lgtm. fix, squash, then merge

File: core-plugins/widgets/Excel-batchsource.json [hunk 10]
```diff
@@ -136,7 +136,7 @@
       "label": "Error Handling",
       "properties": [
         {
-          "widget-type": "select",
+          "widget-type": "hidden",
```
Comment (albertshau) [inline core-plugins/widgets/Excel-batchsource.json:139]: forgot to mention, please remove mention of these from the documentation if they exist

File: core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java (hunk 9 above)
Comment (dli357) [reply core-plugins/src/main/java/io/cdap/plugin/batch/source/ExcelInputReader.java]: Fixed

File: core-plugins/widgets/Excel-batchsource.json (hunk 10 above)
Comment (dli357) [reply core-plugins/widgets/Excel-batchsource.json:139]: Fixed

**Pull request merged**